dowhy/gcm/distribution_change_robust.py

```diff
@@ -0,0 +1,915 @@
+               and above at 1-crop.
+
+        Returns:
+        theta_hat = the point estimate, np.mean(theta_scores) for the scores computed by self.est_scores.
+        std_err = the standard error for theta_hat, sem(theta_scores) for the scores computed by self.est_scores.
+        """
+
+        regressor_kwargs = {} if regressor_kwargs is None else regressor_kwargs
+        regressor_fit_kwargs = {} if regressor_fit_kwargs is None else regressor_fit_kwargs
+        classifier_kwargs = {} if classifier_kwargs is None else classifier_kwargs
+        classifier_fit_kwargs = {} if classifier_fit_kwargs is None else classifier_fit_kwargs
+        calibrator_args = {} if calibrator_args is None else calibrator_args
+        calibrator_fit_kwargs = {} if calibrator_fit_kwargs is None else calibrator_fit_kwargs
+
+        theta_scores = self.est_scores(
+            X_eval,
+            y_eval,
+            T_eval,
+            X_train,
+            y_train,
+            T_train,
+            w_eval=w_eval,
+            w_train=w_train,
+            method=method,  # One of 'regression', 're-weighting', 'MR',
+            regressor=regressor,
+            regressor_args=regressor_args,
+            regressor_kwargs=regressor_kwargs,
+            regressor_fit_kwargs=regressor_fit_kwargs,
+            classifier=classifier,
+            classifier_args=classifier_args,
+            classifier_kwargs=classifier_kwargs,
+            classifier_fit_kwargs=classifier_fit_kwargs,
+            calibrator=calibrator,
+            X_calib=X_calib,
+            T_calib=T_calib,
+            w_calib=w_calib,
+            calibrator_args=calibrator_args,
+            calibrator_kwargs=calibrator_kwargs,
+            calibrator_fit_kwargs=calibrator_fit_kwargs,
+            all_indep=all_indep,
+            crop=crop,
+        )
+
+        if w_eval is not None:
+            w_sort = np.concatenate((w_eval[T_eval == 0], w_eval[T_eval == 1]))  # Order weights in same way as scores
+        else:
+            w_sort = np.ones(np.shape(T_eval))
+        weighted_stats = DescrStatsW(theta_scores, weights=w_sort, ddof=0)
+
+        return weighted_stats.mean, weighted_stats.std_mean
+
+
+def distribution_change_robust(
+    causal_model: ProbabilisticCausalModel,
+    old_data: pd.DataFrame,
+    new_data: pd.DataFrame,
+    target_node: Any,
+    sample_weight=None,
+    xfit=True,
+    xfit_folds=5,
+    train_size=0.5,
+    calib_size=0.0,
+    split_random_state=0,
+    method="MR",  # One of 'regression', 're-weighting', 'MR',
+    regressor=LinearRegression,
+    regressor_args=(),
+    regressor_kwargs=None,
+    regressor_fit_kwargs=None,
+    classifier=LogisticRegression,
+    classifier_args=(),
+    classifier_kwargs=None,
+    classifier_fit_kwargs=None,
+    calibrator=None,
+    calibrator_args=(),
+    calibrator_kwargs=None,
+    calibrator_fit_kwargs=None,
+    all_indep=False,
+    crop=1e-3,
```

Comment:
For consistency with the package, the regressor and classifier should be `PredictionModel` and `ClassificationModel`, respectively. These are gcm interfaces that require a fit and predict method. In this regard, I am wondering if the easier approach here is to simply assume that these models are already implemented with these `args` set. This is currently a lot of parameters to just configure the models. For instance:

```
regressor=LinearRegression,
regressor_args=(),
regressor_kwargs=None,
regressor_fit_kwargs=None
```

can be reduced to just
```
regressor: PredictionModel = create_linear_regressor
```
(with imports from gcm.ml.regression)

Now, if someone wants to have their custom parameters here, they can provide a custom implementation:

```

class MyLinearRegWithFancyParamters(SklearnRegressionModel):
   
   def fit(X, Y):
      mdl = LinearRegression()

      mdl.fit(X, Y, my_param1="..", my_param2=".....")
...

regressor = MyLinearRegWithFancyParamters
```

While this is maybe a bit more cumbersome than via the `args`, it would make the method definition cleaner (i.e., it removes 9 parameters). In practice, I would assume that only a small fraction of users would even use non-default parameters here.

Reply:
I agree that this would be better!

I've pushed a commit cleaning up the args in the regression, classification and calibration. 

However, I'm running into two problems:
1. The `predict` method from `sklearn` regressors returns "flattened" arrays with shape (n_samples,), whereas for your custom class `SklearnRegressionModel` it returns a 2d array with shape (n_samples, 1).
2. Your class `SklearnClassificationModel` method for predicting probabilities is called `predict_probabilities`, whereas for `sklearn` is called `predict_proba`.

Ideally, I would like to keep the syntax exactly as in `sklearn`, in case the user wants to use regression / classification methods that are not in `gcm.ml`. For example, the CPS2015 notebook uses `lightgbm`, which returns flattened predictions for regressions and has the `predict_proba` method.

Reply:
Note that we are using `sklearn` in `dowhy.gcm.ml`. We have a wrapper accordingly: https://github.com/py-why/dowhy/blob/main/dowhy/gcm/ml/regression.py#L28

Here, you can simply use any sklearn model: `prediction_model = lambda: SklearnRegressionModel(HistGradientBoostingRegressor())` (or in your case the `lightgbm` model). Generally, we assume the predictions to be 2D arrays within GCM. In that sense, your code should assume 2D outputs (i.e., it should assume non-flattened outputs). You can, of course, reshape it however you need it, but generally, the prediction model outputs 2D shapes to support multivariate outputs (although this is not really the case yet).

So, in many places, if we need flattened outputs, we are simply doing `my_model.predict(X).reshape(-1)`.

Reply:
To add: For sklearn classification models, it is https://github.com/py-why/dowhy/blob/main/dowhy/gcm/ml/classification.py#L42. 

So, if a user wants to use a sklearn mode, they can just throw them in with these wrapper. In your code, you just expect the outputs according to the GCM interfaces.

Reply:
It seems like the `fit` method for `PredictionModel` does not support sample weights.

Reply:
Oh right, it does not. I assume you necessarily need that? That being said, it would also restrict the type of models that can be used here? Is there a specific sklearn interface for models that support sample weights?

Reply:
We use sample weights in the empirical application. All the sklearn regression and classification models in `gcm.ml` allow sample weights.

Reply:
I double checked, it seems not all `sklearn` models actually support sample weights, which limits the model choices. 

Ok, let's keep it as is for now, since it seems we would need to introduce/modify existing interfaces to properly support this in a consistent way. Regarding typing, if there is a specific interface from `sklearn` that implies a model allowing sample weights, let's use that. If there isn't one, just choose `Any` for now.

Reply:
I'm not sure I follow. If your suggestion is to allow users to pass learners from `sklearn` that allow sample weights directly rather than `PredictionModel`, that won't work because the method for prediction probabilities is `predict_proba` on `sklearn` but `predict_probabilities` in `gcm.ml.classification`.

Reply:
How about having a separate class for sample weights in `gcm.ml.regression`, and similarly for classification?

```
class SklearnRegressionModelWeights(SklearnRegressionModel):
    def fit(self, X: np.ndarray, Y: np.ndarray, sample_weight: np.ndarray = None) -> None:
        self._encoders = auto_fit_encoders(X, Y)
        X = auto_apply_encoders(X, self._encoders)

        self._sklearn_mdl.fit(X=X, y=Y.squeeze(), sample_weight=sample_weight) 
```

Reply:
That would be the best solution.

Reply:
Sounds great! I just pushed an update with the new classes.